mas-foundation/src/main/java/com/ca/mas/core/oauth/CodeVerifierCache.java

```diff
@@ -8,44 +8,45 @@
 
 package com.ca.mas.core.oauth;
 
+import com.ca.mas.core.storage.sharedstorage.SharedPreferencesUtil;
+
 /**
  * Temporary cache to store code verifier
  */
 public class CodeVerifierCache {
 
     private static CodeVerifierCache instance = new CodeVerifierCache();
-
-    private String state;
-    private String codeVerifier;
+    private SharedPreferencesUtil prefUtil = null;
 
     private CodeVerifierCache() {
+        prefUtil = new SharedPreferencesUtil("codeverifier");
     }
 
     public static CodeVerifierCache getInstance() {
         return instance;
     }
 
     public void store(String state, String codeVerifier) {
-        this.state = state;
-        this.codeVerifier = codeVerifier;
+        prefUtil.save("state", state);
```

Comment:
I think, there's a one-one relation between **state** and **codeVerifier**. You should be saving the **codeVerifier** using the **state** as key, Please check it out.

Reply:
@graju256 I also thought to store code verifier using state as key. But we have two overloading methods of take. That is the reason I implemented this way. Any other thoughts here, please let me know.

https://github.com/CAAPIM/Android-MAS-SDK/blob/Stable/mas-foundation/src/main/java/com/ca/mas/foundation/MASAuthCredentialsAuthorizationCode.java

`public class MASAuthCredentialsAuthorizationCode implements MASAuthCredentials {

    private String code;
    private String state;
    private String codeVerifier;

    public MASAuthCredentialsAuthorizationCode(String code, String state) {
        this.code = code;
        this.state = state;
        if (ConfigurationManager.getInstance().isPKCEEnabled()) {
            if (state != null && !state.isEmpty()) {
                this.codeVerifier = CodeVerifierCache.getInstance().take(state);
            } else {
                this.codeVerifier = CodeVerifierCache.getInstance().take();
            }
        }
    }`

Reply:
Still, not convincing to me. 

Method **_take_** with no arguments is simply returning what is there with cache without checking the state.
Whereas, method with argument is strictly checking the state match. Then only it is returning the code-verifier.

Hence, you must use state as key to store the code-verifier.

For backward compatibilty, you can store the latest code-verifier using "__system-default" key